File: programmers/42579베스트앨범/c.cc
#include <string>
#include <vector>
#include<set>
#include<map>

#include<iostream>

using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    map<string,map<int,set<int>>> genresPlayId;
    map<string,int> genresAllPlay;
    set<pair<int,string>> playGenres;
    for(int idx=0; idx<genres.size(); idx++){
        genresPlayId[genres[idx]][plays[idx]].insert(idx);
        genresAllPlay[genres[idx]]+=plays[idx];
    }
    for(auto it=genresAllPlay.begin(); it!=genresAllPlay.end();it++){
        playGenres.insert({it->second, it->first});
    }
    for(auto pg=playGenres.rbegin(); pg!=playGenres.rend(); pg++){
        auto gp = genresPlayId[pg->second];
        auto gpi = gp.rbegin();
        auto no = gpi->second.begin();
        answer.push_back(*no);
        if(gpi->second.size()>1){
            no++;
            answer.push_back(*no);
        }
        else if(gp.size()>1){
            gpi++;
            auto no = gpi->second.begin();
            answer.push_back(*no);
        }
    }
    return answer;
}

File: programmers/42579베스트앨범/c.py
def solution(genres, plays) :
    answer = []
    counting_paly = {}
    music = {}
    for idx in range(len(genres)):
        if genres[idx] in counting_paly:
            counting_paly[genres[idx]] += plays[idx]
            if plays[idx] in music[genres[idx]]:
                music[genres[idx]][plays[idx]].append(idx)
            else:
                music[genres[idx]][plays[idx]] = [idx]
        else:
            counting_paly[genres[idx]] = plays[idx]
            music[genres[idx]] = {plays[idx]:[idx]}
    
    counting_paly = sorted(counting_paly.items(), key=lambda x:x[1], reverse=True)
    for k, v in counting_paly:
        tmp = sorted(music[k].items(), key=lambda x:x[0])
        answer.append(tmp[-1][1].pop(0))
        if(tmp[-1][1]):
            answer.append(tmp[-1][1].pop(0))
        elif(len(tmp)>1):
            answer.append(tmp[-2][1].pop(0))

    return answer